internal/mage/engine.go

```diff
@@ -323,22 +362,79 @@ func devEngineContainer(c *dagger.Client, arches []string) []*dagger.Container {
 			WithFile("/usr/local/bin/"+engineBinName, engineBin(c, arch)).
 			WithDirectory("/usr/local/bin", qemuBins(c, arch)).
 			WithDirectory(engineDefaultStateDir, c.Directory()).
+			WithDirectory("/opt/cni/bin", c.Directory()).
+			WithFile("/opt/cni/bin/dnsname", dnsnameBinary(c, arch)).
+			WithMountedFile("/tmp/cni-plugins.tgz", c.HTTP(cniURL)).
```

Comment:
nit (can all be follow-ups): I forget if `/tmp` is by default a tmpfs, but if not we should delete this file after done in the interest of minimizing the final image size (or just mount a tmpfs at tmp above this line).

Also, if we don't need every plugin in the tarball we could delete the ones we don't need for the same reason. May require combining execs into one to get the layers we want.

Reply:
Done! I pulled the CNI plugin unpacking into a separate step and just copy its output folder over (with just the 4 needed plugins).